test/src/common/bridgePowerController_ut.cpp

```diff
@@ -428,6 +428,173 @@ TEST_F(BridgePowerControllerTest, subsampling2) {
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Time for a bus down cycle
+  M = 1;
+  rtcGetMicroSeconds_fake.return_val = 1713232860000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 6);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 4);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[3], 0);
+
+  // Task waits through the Subsampling off period until the next interval starts
+  curtime += 4 * kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Bus up for next Subsample
+  M = 5;
+  rtcGetMicroSeconds_fake.return_val = 1713233100000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 7);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 5);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[4], 1);
+  curtime += kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Turn off for subsampling
+  M = 6;
+  rtcGetMicroSeconds_fake.return_val = 1713233160000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 8);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 6);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[5], 0);
+  curtime += 4 * kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Bus on
+  M = 10;
+  rtcGetMicroSeconds_fake.return_val = 1713233400000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 9);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 7);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[6], 1);
+  curtime += kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Bus off
+  M = 11;
+  rtcGetMicroSeconds_fake.return_val = 1713233460000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 10);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 8);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[7], 0);
+  curtime += 4 * kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Bus on
+  M = 15;
+  rtcGetMicroSeconds_fake.return_val = 1713233700000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 11);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 9);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[8], 1);
+  curtime += kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Bus off, sleep until end of sample duration
+  M = 16;
+  rtcGetMicroSeconds_fake.return_val = 1713233760000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 12);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 10);
+  EXPECT_EQ(fake_io_write_func_fake.arg1_history[9], 0);
+  curtime += 3 * kOneMinute;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
+
+  // Replicate the bug: wake early
+  M = 18;
+  S = 59;
+  rtcGetMicroSeconds_fake.return_val = 1713233939000000;
+  powerController._update();
+  EXPECT_EQ(fake_io_read_func_fake.call_count, 12);
+  EXPECT_EQ(fake_io_write_func_fake.call_count, 10);
+  curtime += 1000;
+  EXPECT_EQ(xTaskGetTickCount(), curtime);
```

Comment:
Wake a second early. Expect to go back to sleep for a second without touching the bus power. This was where the test failed before I modified the bridge power controller.